robot-server/robot_server/protocols/protocol_store.py

```diff
@@ -203,6 +226,41 @@ def remove(self, protocol_id: str) -> ProtocolResource:
             source=deleted_source,
         )
 
+    def get_usage_info(self) -> List[ProtocolUsageInfo]:
+        """Return information about which protocols are currently being used by runs.
+
+        See the `runs` module for information about runs.
+
+        Results are ordered with the oldest-added protocol first.
+        """
+        select_all_protocol_ids = sqlalchemy.select(protocol_table.c.id).order_by(
+            sqlite_rowid
+        )
+        select_used_protocol_ids = sqlalchemy.select(run_table.c.protocol_id).where(
+            run_table.c.protocol_id.is_not(None)
+        )
+
+        with self._sql_engine.begin() as transaction:
+            all_protocol_ids: List[str] = (
+                transaction.execute(select_all_protocol_ids).scalars().all()
+            )
+            used_protocol_ids: Set[str] = set(
+                transaction.execute(select_used_protocol_ids).scalars().all()
+            )
+
+        # It's probably inefficient to do this processing in Python
+        # instead of as part of the SQL query.
+        # But the number of runs and protocols is on the order of 20, so it's fine.
+        usage_info = [
+            ProtocolUsageInfo(
+                protocol_id=protocol_id,
+                is_used_by_run=(protocol_id in used_protocol_ids),
+            )
+            for protocol_id in all_protocol_ids
+        ]
+
+        return usage_info
```

Comment:
I was playing around with joins, and this can be done in a single query:

```suggestion
        protocols_with_runs = sqlalchemy.join(protocol_table, run_table, isouter=True)
        select_protocol_entries = (
            sqlalchemy.select(protocol_table.c.id, run_table.c.id)
            .select_from(protocols_with_runs)
            .order_by(sqlite_rowid)
        )

        with self._sql_engine.begin() as transaction:
            all_protocol_entries = transaction.execute(select_protocol_entries).all()

        usage_info = [
            ProtocolUsageInfo(
                protocol_id=protocol_id,
                is_used_by_run=run_id is not None,
            )
            for protocol_id, run_id in all_protocol_entries
        ]

        return usage_info
```